domain/cloud/state/state.go

```diff
@@ -0,0 +1,441 @@
+	q := fmt.Sprintf(`
+		DELETE FROM cloud_auth_type
+		WHERE  cloud_uuid = ?
+		AND    auth_type_id NOT IN (%s)`[1:], database.SliceToPlaceholder(authTypeIds))
+
+	args := append([]any{cloudUUID}, transform.Slice(authTypeIds, func(i int) any { return i })...)
+	if _, err := tx.ExecContext(ctx, q, args...); err != nil {
+		return errors.Trace(err)
+	}
+
+	for _, a := range authTypeIds {
+		q := `
+			INSERT INTO cloud_auth_type (uuid, cloud_uuid, auth_type_id)
+			VALUES (?, ?, ?)
+				ON CONFLICT(cloud_uuid, auth_type_id) DO NOTHING`[1:]
+
+		if _, err := tx.ExecContext(ctx, q, utils.MustNewUUID().String(), cloudUUID, a); err != nil {
+			return errors.Trace(err)
+		}
+	}
+	return nil
+}
+
+func updateCACerts(ctx context.Context, tx *sql.Tx, cloudUUID string, certs []string) error {
+	// Delete any existing ca certs - we just delete them all rather
+	// than keeping existing ones as the cert values are long strings.
+	q := `
+		DELETE FROM cloud_ca_cert
+		WHERE  cloud_uuid = ?`
+
+	if _, err := tx.ExecContext(ctx, q, cloudUUID); err != nil {
+		return errors.Trace(err)
+	}
+
+	for _, cert := range certs {
+		q := `
+			INSERT INTO cloud_ca_cert (uuid, cloud_uuid, ca_cert)
+			VALUES (?, ?, ?)`[1:]
+
+		if _, err := tx.ExecContext(ctx, q, utils.MustNewUUID().String(), cloudUUID, cert); err != nil {
+			return errors.Trace(err)
+		}
+	}
+	return nil
+}
+
+func updateRegions(ctx context.Context, tx *sql.Tx, cloudUUID string, regions []cloud.Region) error {
+	regionNames := make([]string, len(regions))
+	for i, r := range regions {
+		regionNames[i] = r.Name
+	}
+
+	// Delete any regions no longer in the list.
+	q := fmt.Sprintf(`
+		DELETE FROM cloud_region
+		WHERE  cloud_uuid = ?
+		AND    name NOT IN (%s)`[1:], database.SliceToPlaceholder(regionNames))
+
+	args := append([]any{cloudUUID}, transform.Slice(regionNames, func(s string) any { return s })...)
+	if _, err := tx.ExecContext(ctx, q, args...); err != nil {
+		return errors.Trace(err)
+	}
+
+	for _, r := range regions {
+		q := `
+			INSERT INTO cloud_region (uuid, cloud_uuid, name, endpoint, identity_endpoint, storage_endpoint)
+			VALUES (?, ?, ?, ?, ?, ?)
+				ON CONFLICT(cloud_uuid, name) DO UPDATE SET name=?, endpoint=?, identity_endpoint=?, storage_endpoint=?`[1:]
+
+		if _, err := tx.ExecContext(ctx, q, utils.MustNewUUID().String(), cloudUUID,
+			// insert
+			r.Name, r.Endpoint, r.IdentityEndpoint, r.StorageEndpoint,
+			// update
+			r.Name, r.Endpoint, r.IdentityEndpoint, r.StorageEndpoint,
+		); err != nil {
+			return errors.Trace(err)
+		}
+	}
+	return nil
+}
+
+func dbCloudFromCloud(tx *sql.Tx, cloudUUID string, cloud cloud.Cloud) (*Cloud, error) {
+	cld := &Cloud{
+		ID:               cloudUUID,
+		Name:             cloud.Name,
+		Endpoint:         cloud.Endpoint,
+		IdentityEndpoint: cloud.IdentityEndpoint,
+		StorageEndpoint:  cloud.StorageEndpoint,
+		SkipTLSVerify:    cloud.SkipTLSVerify,
+	}
+
+	row := tx.QueryRow("SELECT id FROM cloud_type WHERE type = ?", cloud.Type)
```

Comment:
should we `LIMIT 1` here? Otherwise at least we should check that only one row is returned right?

Reply:
Ok sorry, actually `QueryRow` only keeps the first row and discards the rest, so no need for LIMIT.